### PR TITLE
feature: Enable --config-env and --config-file to support multiple configuration environments and files

### PR DESCRIPTION
**Pull request by wchengru**
*Issue #, if available:*
#1615 

*Why is this change necessary?*

*How does it address the issue?*
sam deploy option "--config-env" was not enabled. I enabled this option and allow samconfig.toml to save/load multiple configuration profiles.

*What side effects does this change have?*

*Did you change a dependency in `requirements/base.txt`?*
    *If so, did you run `make update-reproducible-reqs`*

*Checklist:*

- [x] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [x] Write unit tests
- [x] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/cli/cli_config_file.py [hunk 1]
```diff
@@ -31,42 +31,60 @@ class TomlProvider:
...
         :param cmd_names list(str): sam command name as defined by click
         :returns dictionary containing the configuration parameters under specified config_env
         """
 
         resolved_config = {}
 
-        samconfig = SamConfig(config_dir)
+        # Use default sam config file name if config_path only contain the directory
+        samconfig = (
+            SamConfig(os.path.split(config_path)[0], os.path.split(config_path)[1])
```
Comment (sriram-mv) [inline samcli/cli/cli_config_file.py:50]: you can probably use pathlib for this.
pathlib.parent should give you the directory without needing to do splits.

File: samcli/cli/cli_config_file.py [hunk 2]
```diff
@@ -31,42 +31,60 @@ class TomlProvider:
...
+            if config_path
+            else SamConfig(os.getcwd())
+        )
+        # Enable debug level logging by environment variable "SAM_DEBUG"
+        if os.environ.get("SAM_DEBUG", "").lower() == "true":
+            LOG.setLevel(10)
```
Comment (sriram-mv) [inline samcli/cli/cli_config_file.py:56]: What is 10? can we use ENUMs instead?

File: samcli/cli/cli_config_file.py [hunk 3]
```diff
@@ -145,44 +165,88 @@ def configuration_option(*param_decls, **attrs):
...
-        attrs.setdefault("is_eager", True)
-        attrs.setdefault("help", "Read config-env from Configuration File.")
-        attrs.setdefault("expose_value", False)
-        # --config-env is hidden and can potentially be opened up in the future.
-        attrs.setdefault("hidden", True)
-        # explicitly ignore values passed to --config-env, can be opened up in the future.
-        config_env_name = DEFAULT_ENV
+
+    def decorator_configuration_setup(f):
+        configuration_setup_param_devls = ()
```
Comment (sriram-mv) [inline samcli/cli/cli_config_file.py:185]: what does `configuration_setup_param_devls` mean, the meaning is not apparent from the name.

File: samcli/commands/deploy/command.py [hunk 4]
```diff
@@ -42,7 +42,7 @@
     context_settings={"ignore_unknown_options": False, "allow_interspersed_args": True, "allow_extra_args": True},
     help=HELP_TEXT,
 )
-@configuration_option(provider=TomlProvider(section=CONFIG_SECTION))
+@configuration_option(provider=TomlProvider(section=CONFIG_SECTION),)
```
Comment (sriram-mv) [inline samcli/commands/deploy/command.py:45]: is the extra comma required?

File: samcli/lib/config/samconfig.py [hunk 5]
```diff
@@ -106,7 +106,16 @@ def put(self, cmd_names, section, key, value, env=DEFAULT_ENV):
         if not self.document:
             self._read()
             # Empty document prepare the initial structure.
-            self.document.update({env: {self._to_key(cmd_names): {section: {key: value}}}})
+            # self.document is a nested dict, we need to check each layer and add new tables, otherwise duplicated key
+            # in parent layer will override the whole child layer
+            if self.document.get(env, None):
```
Comment (sriram-mv) [inline samcli/lib/config/samconfig.py:111]: is there a testcase to showcase all the cases shown here?

File: tests/unit/commands/deploy/test_guided_context.py [hunk 6]
```diff
@@ -121,3 +123,45 @@ def test_guided_prompts_with_given_capabilities(
             call(f"\t{self.gc.start_bold}Capabilities{self.gc.end_bold}", default=expected_capabilities, type=ANY),
         ]
         self.assertEqual(expected_prompt_calls, patched_prompt.call_args_list)
+
+    @patch("samcli.commands.deploy.guided_context.prompt")
+    @patch("samcli.commands.deploy.guided_context.confirm")
+    @patch("samcli.commands.deploy.guided_context.manage_stack")
+    @patch("samcli.commands.deploy.guided_context.auth_per_resource")
+    @patch("samcli.commands.deploy.guided_context.get_template_data")
+    def test_guided_prompts_check_defaults_public_resources(
```
Comment (sriram-mv) [inline tests/unit/commands/deploy/test_guided_context.py:132]: should the test name be any different?

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/commands/deploy/command.py (hunk 4 above)
Comment (wchengru) [reply samcli/commands/deploy/command.py:45]: I is a typo and have fixed. Thanks for pointing it out.

File: samcli/cli/cli_config_file.py (hunk 3 above)
Comment (wchengru) [reply samcli/cli/cli_config_file.py:185]: It is the parameters of this decorator, as it does not accept any parameters from command line. I put it empty. The name has been changed to "configuration_setup_params".

File: samcli/cli/cli_config_file.py (hunk 2 above)
Comment (wchengru) [reply samcli/cli/cli_config_file.py:56]: Changed it to loggings.DEBUG.

File: samcli/lib/config/samconfig.py (hunk 5 above)
Comment (wchengru) [reply samcli/lib/config/samconfig.py:111]: Yes, they are tested in test_samconfig.py.

File: tests/unit/commands/deploy/test_guided_context.py (hunk 6 above)
Comment (wchengru) [reply tests/unit/commands/deploy/test_guided_context.py:132]: This has been fixed.

File: samcli/cli/cli_config_file.py (hunk 1 above)
Comment (wchengru) [reply samcli/cli/cli_config_file.py:50]: Contributer of SamConfig() is using strings of directory and filename. Here I think to pass in strings would be consistent.

Comment (wchengru) [reply samcli/cli/cli_config_file.py]: I reconsidered this and change to use pathlib to avoid some edge cases.

File: tests/unit/commands/_utils/test_options.py [hunk 7]
```diff
@@ -83,6 +83,15 @@ def test_verify_ctx(self, get_template_data_mock, os_mock):
         self.assertEqual(ctx.template_dict, "dummy_template_dict")
         os_mock.path.abspath.assert_called_with(expected)
 
+    def test_verify_ctx_template_file_param(self):
```
Comment (sriram-mv) [inline tests/unit/commands/_utils/test_options.py:87]: can we add a specific test for this in test_samconfig.py file

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: tests/unit/commands/_utils/test_options.py (hunk 7 above)
Comment (wchengru) [reply tests/unit/commands/_utils/test_options.py:87]: sure, new test case is added in that file

**Comment by faboulaye** [comment]
When this issue will be available ?

File: tests/unit/cli/test_cli_config_file.py [hunk 8]
```diff
@@ -26,33 +27,41 @@ def setUp(self):
...
         self.assertEqual(self.toml_provider(config_dir, self.config_env, [self.cmd_name]), {})
 
+    def test_toml_invalid_file_name(self):
+        config_dir = tempfile.gettempdir()
+        config_path = Path(config_dir, "mysamconfig.toml")
+        config_path.write_text("version=0.1\n[config_env.topic.parameters]\nword='clarity'\n")
+        config_path_invalid = Path(config_dir, "samconfig.toml")
+
+        self.assertEqual(self.toml_provider(config_path_invalid, self.config_env, [self.cmd_name]), {})
+
```
Comment (awood45) [inline tests/unit/cli/test_cli_config_file.py:65]: Can we add a test with an absolute path? May need to have another function call generate it so it works on whatever machine runs this test, but I didn't notice that case being covered.

File: tests/unit/commands/samconfig/test_samconfig.py [hunk 9]
```diff
@@ -714,6 +718,27 @@ def test_override_with_cli_params_and_envvars(self, do_cli_mock):
                 {"A": "123", "C": "D", "E": "F12!", "G": "H"},
             )
 
+    @patch("samcli.commands.validate.validate.do_cli")
+    def test_secondary_option_name_template_validate(self, do_cli_mock):
+        # "--tempalte" is an alias of "--template-file"
```
Comment (awood45) [inline tests/unit/commands/samconfig/test_samconfig.py:723]: s/tempalte/template

**Review by awood45 — APPROVED** [review]
Looks good, ideally we can add one more test and there's a comment that has a typo.

File: tests/unit/commands/samconfig/test_samconfig.py (hunk 9 above)
Comment (wchengru) [reply tests/unit/commands/samconfig/test_samconfig.py:723]: Will fix this in later PRs.

File: tests/unit/cli/test_cli_config_file.py (hunk 8 above)
Comment (wchengru) [reply tests/unit/cli/test_cli_config_file.py:65]: I think we have a temp directory generated by `config_dir = tempfile.gettempdir()`, and that will be the absolute path.

**Review by sriram-mv — APPROVED** [review]
